src/StorageEntityManager.sol

```diff
@@ -0,0 +1,185 @@
+        storageEntity.storageProviders = storageProviders;
+        storageEntity.isActive = true;
+
+        Storage.s().entityAddresses.push(owner);
+
+        // Mark each storage provider as used
+        for (uint i = 0; i < storageProviders.length; i++) {
+            Storage.s().usedStorageProviders[storageProviders[i]] = true;
+        }
+
+        emit StorageEntityCreated(msg.sender, owner, storageProviders);
+    }
+
+    function addStorageProviders(
+        address owner,
+        uint64[] calldata storageProviders
+    ) public onlyStorageEntity(owner) {
+        _checkIfAnyStorageProviderIsUsed(storageProviders);
+
+        Storage.StorageEntity storage se = Storage.s().storageEntities[
+            owner
+        ];
+
+        _checkIfStorageEntityExists(se);
+
+        for (uint i = 0; i < storageProviders.length; i++) {
+            se.storageProviders.push(storageProviders[i]);
+            Storage.s().usedStorageProviders[storageProviders[i]] = true;
+        }
+
+        emit StrorageProvidersAdded(
+            msg.sender,
+            owner,
+            storageProviders
+        );
+    }
+
+    function removeStorageProviders(
+        address owner,
+        uint64[] calldata storageProviders
+    ) public onlyStorageEntity(owner) {
+        Storage.StorageEntity storage se = Storage.s().storageEntities[
+            owner
+        ];
+
+        _checkIfStorageEntityExists(se);
+
+        for (uint j = 0; j < storageProviders.length; j++) {
+            uint64 sp = storageProviders[j];
+            Storage.s().usedStorageProviders[sp] = false;
+            for (uint i = 0; i < se.storageProviders.length; i++) {
+                if (se.storageProviders[i] == sp) {
+                    se.storageProviders[i] = se.storageProviders[
+                        se.storageProviders.length - 1
+                    ];
+                    se.storageProviders.pop();
+                    Storage.s().usedStorageProviders[sp] = false;
+                    break;
+                }
+            }
+        }
+
+        emit StorageProviderRemoved(msg.sender, owner, storageProviders);
+    }
+
+    function changeStorageEntityActiveStatus(
+        address owner,
+        bool isActive
+    ) public onlyStorageEntity(owner) {
+        Storage.StorageEntity storage se = Storage.s().storageEntities[
+            owner
+        ];
+
+        _checkIfStorageEntityExists(se);
+
+        se.isActive = isActive;
+
+        emit StorageEntityActiveStatusChanged(msg.sender, owner, isActive);
+    }
+
+    function _checkIfAnyStorageProviderIsUsed(
+        uint64[] calldata storageProviders
+    ) internal view {
+        for (uint i = 0; i < storageProviders.length; i++) {
+            if (Storage.s().usedStorageProviders[storageProviders[i]]) {
+                revert Errors.StorageProviderAlreadyUsed();
+            }
+        }
+    }
+
+    function _checkIfStorageEntityExists(Storage.StorageEntity storage se)
```

Comment:
```suggestion
    function _ensureStorageEntityNotExists(Storage.StorageEntity storage se)
```
That way we're more explicit that this reverts. Same with `_checkIfAnyStorageProviderIsUsed`